IOOBE when inserting a number of documents with heap byte buffers

Fail witnessed on `single-typed-value-to-nils-across-blocks`:
```
:result
   #xt/error [:fault :xtdb.error/unknown "Index 1 out of bounds for length 1" {:sql "SELECT * FROM docs ORDER BY _id"}],
  :result-data
  {:clojure.test.check.properties/error
    #xt/error [:fault :xtdb.error/unknown "Index 1 out of bounds for length 1" {:sql "SELECT * FROM docs ORDER BY _id"}]},
  :time-shrinking-ms 3225,
  :smallest
  [{:vec-name "0",
    :vs
    [#object[java.nio.HeapByteBuffer 0x74405179 "java.nio.HeapByteBuffer[pos=0 lim=1 cap=1]"]],
    :type
    {:gen
     #object[clojure.test.check.generators$gen_bind$fn__4343 0x34ac6d65 "clojure.test.check.generators$gen_bind$fn__4343@34ac6d65"]}}]},
```

In this instance, running the contents of the property test with `smallest` wasn't particularly enlightening - not knowing the underlying byte array doesn't help matters.

As "minimal" as possible to throw an error:
```
(t/deftest minimal-ioobe-repro
  (let [node (xtn/start-node {:log [:in-memory {:instant-src (tu/->mock-clock)}]
                              :compactor {:threads 0}})
        vals-to-gen 1000
        varbinary-vals (gen/sample tg/varbinary-gen vals-to-gen)]
    (xt/execute-tx node [(into [:put-docs :docs] (map-indexed (fn [i v] {:xt/id (+ 1000 i) :field v}) varbinary-vals))])
    (t/is (= 2 (count (xt/q node "FROM xt.txs"))))
    (t/is (= 1000 (count (xt/q node "SELECT * FROM docs ORDER BY _id"))))))
```

From the stacktrace, I reckon it's an error we throw when returning results out of postgres, similar to recent string escaping bugs:

```
Caused by: xtdb.error.Fault: Index 94 out of bounds for length 94
	at app//xtdb.error$fault.invokeStatic(error.clj:90)
	at app//xtdb.error$fault.invoke(error.clj:82)
	at app//xtdb.error$fn__2187.invokeStatic(error.clj:204)
	at app//xtdb.error$fn__2187.invoke(error.clj:173)
	at app//xtdb.error$fn__2124$G__2119__2131.invoke(error.clj:14)
	at app//xtdb.api$plan_q$reify__4270$fn__4271.invoke(api.clj:120)
	at app//xtdb.api$with_conn.invokeStatic(api.clj:58)
	at app//xtdb.api$with_conn.invoke(api.clj:41)
	at app//xtdb.api$plan_q$reify__4270.reduce(api.clj:116)
	at app//clojure.core$reduce.invokeStatic(core.clj:6964)
	at app//clojure.core$into.invokeStatic(core.clj:7038)
	at app//clojure.core$into.invoke(core.clj:7029)
	at app//xtdb.api$q.invokeStatic(api.clj:168)
	at app//xtdb.api$q.invoke(api.clj:128)
	at app//xtdb.api$q.invokeStatic(api.clj:165)
	at app//xtdb.api$q.invoke(api.clj:128)
	... 32 more
Caused by: java.lang.ArrayIndexOutOfBoundsException: Index 94 out of bounds for length 94
	at org.postgresql.util.PGbytea.toBytesOctalEscaped(PGbytea.java:112)
	at org.postgresql.util.PGbytea.toBytes(PGbytea.java:55)
	at org.postgresql.jdbc.PgResultSet.getBytes(PgResultSet.java:2953)
	at org.postgresql.jdbc.PgResultSet.internalGetObject(PgResultSet.java:255)
	at org.postgresql.jdbc.PgResultSet.getObject(PgResultSet.java:3184)
	at xtdb.jdbc.XtConnection$XtResultSet.getObject(XtConnection.kt:55)
	at xtdb.next.jdbc$read_col.invokeStatic(jdbc.clj:13)
	at xtdb.next.jdbc$read_col.invokePrim(jdbc.clj)
	at xtdb.next.jdbc$builder_fn$reify__3970.with_column(jdbc.clj:43)
	at next.jdbc.result_set$row_builder$fn__1648.invoke(result_set.clj:447)
	at clojure.lang.LongRange.reduce(LongRange.java:170)
	at clojure.core$reduce.invokeStatic(core.clj:6964)
	at clojure.core$reduce.invoke(core.clj:6947)
	at next.jdbc.result_set$row_builder.invokeStatic(result_set.clj:447)
	at next.jdbc.result_set$row_builder.invoke(result_set.clj:443)
	at next.jdbc.result_set$mapify_result_set$reify__1709.iterator(result_set.clj:519)
	at clojure.core.protocols$iter_reduce.invokeStatic(protocols.clj:52)
	at clojure.core.protocols$fn__8260.invokeStatic(protocols.clj:74)
	at clojure.core.protocols$fn__8260.invoke(protocols.clj:74)
	at clojure.core.protocols$fn__8207$G__8202__8220.invoke(protocols.clj:13)
	at clojure.core$reduce.invokeStatic(core.clj:6965)
	at clojure.core$into.invokeStatic(core.clj:7038)
	at clojure.core$into.invoke(core.clj:7029)
	at xtdb.api$plan_q$reify__4270$fn__4271$fn__4272.invoke(api.clj:124)
	at clojure.core$map$fn__5952$fn__5953.invoke(core.clj:2759)
	at next.jdbc.result_set$reduce_result_set.invokeStatic(result_set.clj:709)
	at next.jdbc.result_set$reduce_result_set.invoke(result_set.clj:701)
	at next.jdbc.result_set$reduce_stmt.invokeStatic(result_set.clj:725)
	at next.jdbc.result_set$reduce_stmt.invoke(result_set.clj:715)
	at next.jdbc.result_set$fn$reify__1772.reduce(result_set.clj:872)
	at clojure.core$transduce.invokeStatic(core.clj:7025)
	at clojure.core$transduce.invoke(core.clj:7012)
	at xtdb.api$plan_q$reify__4270$fn__4271.invoke(api.clj:124)
	... 42 more
```

--- a/core/src/main/clojure/xtdb/pgwire/types.clj
+++ b/core/src/main/clojure/xtdb/pgwire/types.clj
@@ -601,14 +601,22 @@
                :typcategory :user-defined
                :typsend "byteasend"
                :typreceive "bytearecv"
-               :read-text (fn [_env ba] ba)
+               ;; Use PostgreSQL hex format (\x<hex>) for text encoding to avoid escape format issues.
+               ;; See: https://www.postgresql.org/docs/current/datatype-binary.html
+               :read-text (fn [_env ba]
+                            ;; Decode hex format from client parameters
+                            (let [s (read-utf8 ba)]
+                              (if (str/starts-with? s "\\x")
+                                (Hex/decodeHex (.substring s 2))
+                                ba)))
                :read-binary (fn [_env ba] (ByteBuffer/wrap ba)
                               (throw (UnsupportedOperationException.)))
                :write-text (fn [_env ^VectorReader rdr idx]
                              (let [bb (.getBytes rdr idx)
                                    ba ^bytes (byte-array (.remaining bb))]
                                (.get bb ba)
-                               ba))
+                               ;; Encode as hex format for client results
+                               (utf8 (str "\\x" (Hex/encodeHexString ba)))))
                :write-binary (fn [_env ^VectorReader rdr idx]
                                (let [bb (.getBytes rdr idx)
                                      ba ^bytes (byte-array (.remaining bb))]

--- a/src/main/clojure/xtdb/test_generators.clj
+++ b/src/main/clojure/xtdb/test_generators.clj
@@ -61,12 +61,11 @@
 (def duration-gen (gen/return #xt/duration "PT1S"))
 (def interval-gen (gen/return #xt/interval "P1YT1S"))
 
-;; Exclude varbinary generators due to issue #4793
 (def simple-type-gens
   [nil-gen bool-gen
    i8-gen i16-gen i32-gen i64-gen
    f64-gen decimal-gen
-   utf8-gen #_varbinary-gen
+   utf8-gen varbinary-gen
    keyword-gen uuid-gen uri-gen
    instant-gen local-date-gen local-time-gen
    local-datetime-gen offset-datetime-gen zoned-datetime-gen

--- a/src/test/clojure/xtdb/pgwire_test.clj
+++ b/src/test/clojure/xtdb/pgwire_test.clj
@@ -2283,6 +2283,11 @@ ORDER BY t.oid DESC LIMIT 1"
       (t/is (= [{:v "dollar $quoted$ string"}] (q conn ["SELECT $$dollar $quoted$ string$$ AS v"]))
             "dollar quoted string"))))
 
+(deftest test-varbinary-escaped-4793
+  (let [^bytes ba (byte-array [92])]
+    (with-open [conn (jdbc-conn)]
+      (t/is (Arrays/equals ba ^bytes (:v (first (jdbc/execute! conn ["SELECT X('5c') AS v"]))))
+            "reading varbinary result"))))
 
 (t/deftest test-patch
   (with-open [conn (jdbc-conn)]

